### PR TITLE
macOS: fix window resizing regression.

### PR DESCRIPTION
**Pull request by prideout**

### CODE REVIEW

**Comment by prideout** [comment]
@bejado do you think I need to remove the observer somehow during `terminate`?  Unsure of how to do that.

**Comment by bejado** [comment]
@prideout Yes, I think it's a good idea to remove the observer. `addObserverForName` will return an opaque observer object which you can pass to [`removeObserver`](https://developer.apple.com/documentation/foundation/nsnotificationcenter/1413994-removeobserver?language=objc).

Also, I'm wondering why we didn't need this before? Is this another Catalina issue?

**Comment by prideout** [comment]
Ah thanks, will try that!  Regarding why we didn't need this until now: before our catalina fix we simply called `update` on every frame (which was overkill).  After our Catalina fix we made it so it only happens when the context was assigned to a new view; since it occurs in another thread we needed to prevent over-synchronization.

**Comment by pixelflinger** [comment]
I don't think this will work if we had several windows; because it would call update() on the context at some random time -- maybe when the context is bound to the 'other' window.

Also, don't we have any locking issue here? This update() method will now be called on the UI thread at any time relative to the GL thread. Can't that cause problems?
I wasn't worried about this with the 'sync' method because at least we knew where the gl thread was.

Are we sure that Catalina doesn't support calling update() from the gl thread?

All that said... I can't think of a better solution though.

**Comment by prideout** [comment]
When I try calling update from the GL thread, I see EXC_BAD_INSTRUCTION.

**Comment by prideout** [comment]
@pixelflinger PTAL, I added an if check to the event handler and added cautionary comments for our users.

File: filament/backend/src/opengl/PlatformCocoaGL.mm [hunk 1]
```diff
@@ -76,8 +76,10 @@
         [NSNotificationCenter defaultCenter] addObserverForName: NSViewGlobalFrameDidChangeNotification
         object: nil
         queue: nil
-        usingBlock: ^ (NSNotification * note) {
-            [nsOpenGLContext update];
+        usingBlock: ^ (NSNotification* note) {
+            if (note.object == pImpl->mCurrentView) {
```
Comment (prideout) [inline filament/backend/src/opengl/PlatformCocoaGL.mm]: Dot instead of arrow because Objective C.  :)

Comment (prideout) [reply filament/backend/src/opengl/PlatformCocoaGL.mm]: Wait this is potentially the wrong mCurrentView, let me tweak this by adding a copy to the closure...

Comment (prideout) [reply filament/backend/src/opengl/PlatformCocoaGL.mm]: Actually that's somewhat non-trivial, I'll need to move the observer creation into makeCurrent 😭

Comment (prideout) [reply filament/backend/src/opengl/PlatformCocoaGL.mm]: Fixed this up.  Should be ready for review now.

File: filament/backend/src/opengl/PlatformCocoaGL.mm [hunk 2]
```diff
@@ -36,6 +36,7 @@
 struct PlatformCocoaGLImpl {
     NSOpenGLContext* mGLContext = nullptr;
     NSView* mCurrentView = nullptr;
+    NSObject* mObserver = nullptr;
```
Comment (bejado) [inline filament/backend/src/opengl/PlatformCocoaGL.mm]: Technically `addObserverForName` returns an `id<NSObject>`

File: filament/backend/src/opengl/PlatformCocoaGL.mm [hunk 3]
```diff
@@ -74,6 +75,9 @@
 }
 
 void PlatformCocoaGL::terminate() noexcept {
+    if (pImpl->mObserver) {
```
Comment (bejado) [inline filament/backend/src/opengl/PlatformCocoaGL.mm]: I don't think we need to check if `mObserver` isn't `nil`. If it is, then `removeObserver:nil` will simply do nothing. Same with down below.

File: filament/backend/src/opengl/PlatformCocoaGL.mm (hunk 2 above)
Comment (prideout) [reply filament/backend/src/opengl/PlatformCocoaGL.mm]: Ah thanks!

File: filament/backend/src/opengl/PlatformCocoaGL.mm (hunk 3 above)
Comment (prideout) [reply filament/backend/src/opengl/PlatformCocoaGL.mm]: I was wondering about that... thanks, fixed it up!

**Pull request merged**